src/dune_file_watcher/dune_file_watcher.ml

```diff
@@ -306,25 +310,34 @@ module Buffer = struct
          List.rev !lines)
 end
 
-let special_file_for_fs_sync =
-  let path =
-    lazy
-      (let dir = Path.Build.relative Path.Build.root ".sync" in
-       Path.Build.relative dir "token")
-  in
-  fun () -> Lazy.force path
-
-let special_file_for_inotify_sync_absolute =
-  lazy (Path.to_absolute_filename (Path.build (special_file_for_fs_sync ())))
-
-let is_special_file_for_inotify_sync (path : Path.t) =
-  match path with
-  | In_source_tree _ -> false
-  | External _ ->
-    String.equal (Path.to_string path)
-      (Lazy.force special_file_for_inotify_sync_absolute)
-  | In_build_dir build_path ->
-    Path.Build.( = ) build_path (special_file_for_fs_sync ())
+let special_dir_for_fs_sync =
+  lazy
+    (Path.Build.relative Path.Build.root ".sync" |> Path.build |> Path.to_string)
+
+let emit_sync t =
+  let id = Sync_id.gen () in
+  let fn = id |> Sync_id.to_int |> string_of_int in
+  let path = Filename.concat (Lazy.force special_dir_for_fs_sync) fn in
+  Unix.close (Unix.openfile path [ O_WRONLY; O_CREAT; O_TRUNC ] 0o666);
+  Table.set t.sync_table fn id;
+  id
+
+let is_special_file_for_fs_sync ~path_as_reported_by_file_watcher =
+  (* We use string matching here and that's fine because we match on the path
+     reported by the file watcher backend which will always report the same
+     string that was used to setup the watch. *)
+  Filename.dirname path_as_reported_by_file_watcher
+  = Lazy.force special_dir_for_fs_sync
+
+let consume_sync_event table path =
+  let basename = Filename.basename path in
+  match Table.find table basename with
+  | None -> None
+  | Some id ->
+    (try Unix.unlink path with
+    | _ -> ());
```

Comment:
We have `Fpath.unlink_no_err` which is a little safer.

Reply:
Thanks, updated